Fix errors in `control_flow` notebook

edoardob90 opened the pull request:
Fix #234

edoardob90 commented:
Except the first point (which is not very clear to me), I think I addressed all the issues of #234

Snowwpanda commented on control_flow.ipynb:
```diff
@@ -972,7 +1212,116 @@
+    "Simply using `raise` without arguments is useful when:\n",
+    "\n",
+    "- We want to preserve the original traceback (i.e., the full *error context*)\n",
+    "- We want to do something if the error occurs, but we still want it to propagate up and be caught by the intepreter (or another `try-except` block) "
```
```suggestion
    "- We want to do something if the error occurs, but we still want it to propagate up and be caught by the interpreter (or another `try-except` block) "
```

Snowwpanda commented on control_flow.ipynb:
```diff
@@ -972,7 +1212,116 @@
+    "A few advice to keep in mind when using `raise`:\n",
+    "\n",
+    "1. Always raise specific exceptions rather than generic ones\n",
+    "2. Provide clear and detailes error messages\n",
```
```suggestion
    "2. Provide clear and detailed error messages\n",
```

Snowwpanda commented on control_flow.ipynb:
```diff
@@ -802,7 +969,80 @@
+   "source": [
+    "### Tips to improve performance\n",
+    "\n",
+    "These tips are most likely out of scope for an introductory tutorial, but for the sake of completess, you can investigate if any of these is applicable to your case:\n",
```
```suggestion
    "These tips are most likely out of scope for an introductory tutorial, but for the sake of completeness, you can investigate if any of these is applicable to your case:\n",
```

Snowwpanda commented on tutorial/tests/test_control_flow.py:
```diff
@@ -156,20 +175,39 @@ def reference_find_pair(nums: List[int]) -> int:
+            for j in range(i + 1, n - 1):
+                for k in range(j + 1, n):
+                    if nums[i] + nums[j] + nums[k] == 2020:
+                        return nums[i] * nums_2[j] * nums[k]
```
```suggestion
                        return nums[i] * nums[j] * nums[k]
```
typo

Snowwpanda reviewed (commented):
Nice, changed quite a lot, i like the examples you are giving, checked all the exercises once, seems to work for me. Just some typos left.

edoardob90 commented:
Thanks @Snowwpanda! I re-read it twice but didn't catch those typos. I'll make a few small changes (and remove unused files), then we should merge.

despadam commented:
> Except the first point (which is not very clear to me), I think I addressed all the issues of #234

The first point means that we are missing explanations on more advanced syntax. For example in the first theoretical part:
`Conditionals` explains the basic `if-elif-else` block, and then in the `Quiz on conditionals` Q3, Q4, Q5 will ask questions on theory that we have not yet explained, i.e. one-line syntax, ternary operator and multiple conditions in a single statement.

edoardob90 commented:
> > Except the first point (which is not very clear to me), I think I addressed all the issues of #234
> 
> The first point means that we are missing explanations on more advanced syntax. For example in the first theoretical part: `Conditionals` explains the basic `if-elif-else` block, and then in the `Quiz on conditionals` Q3, Q4, Q5 will ask questions on theory that we have not yet explained, i.e. one-line syntax, ternary operator and multiple conditions in a single statement.

Ok, got it. I'm finishing with the Codespaces setup, then I'll fix the last things here.

edoardob90 commented:
> > > Except the first point (which is not very clear to me), I think I addressed all the issues of #234
> > 
> > 
> > The first point means that we are missing explanations on more advanced syntax. For example in the first theoretical part: `Conditionals` explains the basic `if-elif-else` block, and then in the `Quiz on conditionals` Q3, Q4, Q5 will ask questions on theory that we have not yet explained, i.e. one-line syntax, ternary operator and multiple conditions in a single statement.
> 
> Ok, got it. I'm finishing with the Codespaces setup, then I'll fix the last things here.

I wrote a bit more about "Conditionals", including some practical examples. There're a lot of `def` and functions even though they are covered in the next notebook, but there's no way to avoid, I think.

Also: the Q5 of the first quiz is okay because they know about logical expressions and operators from the built-in datatypes.

despadam reviewed (changes requested):
Two more comments:
1. Can we expand all cells? At least in the theoretical parts, just to be sure the users will not miss anything. For the longer chilly pepper exercises, I understand it if you still want to hide them.
2. Add some explanation for one-line syntax with iterables. For example, the warm-up exercise `1. Write a Python program that returns the characters in a string and their indexes` suggest this reference solution:
```
return [(char, index) for index, char in enumerate(string)]
```
but I believe a lot of people at this point would just write it as:
```
result = []
for index, char in enumerate(string):
    result.append((char, index))
return result
```
In fact, quite a few people last time were asking how to get from the expanded solution to the compact one, and it is something that it is not clearly explained in the notebook.

edoardob90 commented:
> Two more comments:
> 
> 1. Can we expand all cells? At least in the theoretical parts, just to be sure the users will not miss anything. For the longer chilly pepper exercises, I understand it if you still want to hide them.
> 2. Add some explanation for one-line syntax with iterables. For example, the warm-up exercise `1. Write a Python program that returns the characters in a string and their indexes` suggest this reference solution:
> 
> ```
> return [(char, index) for index, char in enumerate(string)]
> ```
> 
> but I believe a lot of people at this point would just write it as:
> 
> ```
> result = []
> for index, char in enumerate(string):
>     result.append((char, index))
> return result
> ```
> 
> In fact, quite a few people last time were asking how to get from the expanded solution to the compact one, and it is something that it is not clearly explained in the notebook.

1. Sure, I didn't remember that the notebook's metadata store the "status" of each cell (expanded/collpased)
2. I'll change the reference solution because comprehension expressions are discussed in "Functional Programming"

despadam commented:
Thanks! The fix to replace the reference solution, instead of using comprehension expressions, should apply to all exercises.

despadam commented:
Last comment (I promise) would be to also add some simple exercises in the `alternating loops` section, perhaps 1 for each way.

Snowwpanda commented on control_flow.ipynb:
```diff
@@ -798,6 +796,59 @@
+    "\n",
+    "You must use an `if-else` inside a loop.\n",
+    "\n",
+    "Example: `[0, 3, 1, 2]` should return `[3]` because:\n",
```
Just a detail but I would like the example to be [0, 4, 1, 3]` should return `[4] to highlight you must be strictly larger.

edoardob90 replied:
I think it's already written in the text: "keeping only numbers larger than their position in the list"

Snowwpanda commented on tutorial/tests/test_control_flow.py:
```diff
@@ -68,9 +67,13 @@ def test_range_of_nums(start: int, end: int, function_to_test) -> None:
 
 
-def reference_sqrt_of_nums(nums: list[int]) -> list[float]:
+def reference_sqrt_of_nums(nums: List[int]) -> List[float]:
```
Make input variable name consistent, solution calls this numbers

Snowwpanda reviewed (changes requested):
Minor changes.
I like having examples with the loops!

despadam reviewed (changes requested):
Overall I like the examples, I just thing they need some clarifications or that we can also use them as examples to solve together with the students

edoardob90 commented on tutorial/tests/test_control_flow.py:
```diff
@@ -123,6 +123,99 @@ def test_divide_until(num: int, function_to_test) -> None:
+
+def reference_filter_by_position(numbers: List[int]) -> List[int]:
+    result = set()
+    for pos, number in enumerate(numbers, start=1):
```
That's easy: remove `start=1` and do `i + 1`, although it's uglier 😁

edoardob90 replied:
Better: I added a paragraph when talking about `enumerate`

edoardob90 commented on control_flow.ipynb:
```diff
@@ -798,6 +796,59 @@
+   "source": [
+    "#### Exercise: conditionals inside loops\n",
+    "\n",
+    "Complete the function `solution_filter_by_position` below that filters a list of integers, keeping only numbers larger than their position in the list (**1-based index**).\n",
```
I added a paragraph when talking about `enumerate`

edoardob90 commented on tutorial/tests/test_control_flow.py:
```diff
@@ -123,6 +123,99 @@ def test_divide_until(num: int, function_to_test) -> None:
+    for char in text:
+        if char not in "1234567890":
+            return False
+    else:
```
Do you mean specifically the `for-else` construct? That's tough, I guess we can only use the `ast` module for that.
